### PR TITLE
lightning: fix data lost when partial write and server is busy

### LINKED ISSUE #43768: lightning will skip importing data when "partial write" + "server is busy"
close https://github.com/pingcap/tiflash/issues/7340

In https://github.com/pingcap/tidb/pull/42780 I changed the logic for "partial write", which means lightning should write partial data of a job to TiKV, ingest them, and then write rest data, ingest, ...

```golang
	for {
		err := local.writeToTiKV(ctx, job)
		if err != nil {
			if !local.isRetryableImportTiKVError(err) {
				return err
			}
			// if it's retryable error, we retry from scanning region
			log.FromContext(ctx).Warn("meet retryable error when writing to TiKV",
				log.ShortError(err), zap.Stringer("job stage", job.stage))
			job.convertStageTo(needRescan)
			job.lastRetryableErr = err
			return nil
		}

		err = local.ingest(ctx, job)
		if err != nil {
			if !local.isRetryableImportTiKVError(err) {
				return err
			}
			log.FromContext(ctx).Warn("meet retryable error when ingesting",
				log.ShortError(err), zap.Stringer("job stage", job.stage))
			job.lastRetryableErr = err
			return nil
		}

		if job.writeResult == nil || job.writeResult.remainingStartKey == nil {
			return nil
		}
		job.keyRange.start = job.writeResult.remainingStartKey
		job.convertStageTo(regionScanned)
	}
```

However, above code introduce https://github.com/pingcap/tidb/issues/43363 and this bug 😢 

`writeToTiKV` and `ingest` has two output, one is the returned `err`, the other is `job.stage`. If we want to continue the for-loop to write rest data after a "partial write", we should only allow `err == nil && job.stage == ingested`. This bug is caused by we continue for `err == nil && job.stage == wrote`, where `wrote` means the job should be retried to ingest later.

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #43768

Problem Summary:

### What is changed and how it works?

please see the linked issue

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local_test.go [hunk 1]
```diff
@@ -1452,6 +1462,125 @@ func TestPartialWriteIngestErrorWillPanic(t *testing.T) {
 	require.Equal(t, []uint64{1}, apiInvokeRecorder["MultiIngest"])
 }
 
+func TestPartialWriteIngestBusy(t *testing.T) {
+	log.InitLogger(&log.Config{Level: "debug"}, "debug")
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// let lightning meet any error that will call convertStageTo(needRescan)
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local_test.go:1470]:
```suggestion
```

File: br/pkg/lightning/backend/local/local_test.go [hunk 2]
```diff
@@ -1452,6 +1462,125 @@ func TestPartialWriteIngestErrorWillPanic(t *testing.T) {
 	require.Equal(t, []uint64{1}, apiInvokeRecorder["MultiIngest"])
 }
 
+func TestPartialWriteIngestBusy(t *testing.T) {
+	log.InitLogger(&log.Config{Level: "debug"}, "debug")
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local_test.go:1466]:
```suggestion
```

File: br/pkg/lightning/backend/local/local_test.go [hunk 3]
```diff
@@ -1452,6 +1462,125 @@ func TestPartialWriteIngestErrorWillPanic(t *testing.T) {
...
+		defer wg.Done()
+		err := local.startWorker(ctx, jobCh, jobOutCh, &jobWg)
+		require.NoError(t, err)
+	}()
+
+	jobWg.Wait()
+	cancel()
+	wg.Wait()
+
+	require.Equal(t, int64(2), f.importedKVCount.Load())
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/local_test.go:1578]: here is the check

**Review by gozssky — APPROVED** [review]

**Review by D3Hunter — APPROVED** [review]

File: br/pkg/lightning/backend/local/local.go [hunk 4]
```diff
@@ -1364,7 +1364,7 @@ func (local *Backend) executeJob(
 			job.lastRetryableErr = err
 			return nil
 		}
-		if job.stage == needRescan {
+		if job.stage != ingested {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/local.go:1367]: add some comment below to explain why we continue on `ingested` state?

**Pull request merged**